frontend/src/components/dashboard/JoinOrganizationModal.tsx

```diff
@@ -0,0 +1,437 @@
+import Modal from "@components/_shared/Modal";
+import { DefaultTooltip } from "@components/ui/tooltip";
+import { Button } from "@components/ui/button";
+import { api } from "@utils/api";
+import Spinner from "@components/_shared/Spinner";
+import { ErrorAlert } from "@components/_shared/Alerts";
+import { z } from "zod";
+import { zodResolver } from "@hookform/resolvers/zod";
+import {
+  OrganizationFormType,
+  OrganizationSchema,
+} from "@schema/organization.schema";
+import { Form, FormField } from "@components/ui/form";
+import { useForm, UseFormReturn } from "react-hook-form";
+import TextDivisor from "@components/_shared/TextDivisor";
+import TextEditor from "@components/_shared/TextEditor";
+import { OnboardingFormType } from "@schema/onboarding.schema";
+import { toast } from "@components/ui/use-toast";
+
+export default function JoinOrganizationModalButton({
+  render,
+}: {
+  render?: (setIsShow: Dispatch<SetStateAction<boolean>>) => React.ReactNode;
+}) {
+  const [isShow, setIsShow] = useState<boolean>(false);
+
+  if (render) {
+    return (
+      <>
+        {render(setIsShow)}
+        <JoinOrganizationModal isShow={isShow} setIsShow={setIsShow} />
+      </>
+    );
+  }
+
+  return (
+    <>
+      <DefaultTooltip content="Request to join an organisation or the creation of a new one">
+        <Button
+          variant="secondary"
+          className="justify-between gap-2"
+          onClick={() => setIsShow(true)}
+        >
+          <BuildingIcon className="h-4 w-4" />
+          Join Organisation
+        </Button>
+      </DefaultTooltip>
+      <JoinOrganizationModal isShow={isShow} setIsShow={setIsShow} />
+    </>
+  );
+}
+
+const rte = z
+  .string()
+  .min(1 + "<p></p>".length, {
+    message: `Text must be at least 1 character long.`,
+  })
+  .refine((v) => v != "<p></p>");
+
+const joinOrganizationSchema = z
+  .object({
+    isCreation: z.literal(false),
+    org: OrganizationSchema,
+    message: rte,
+    confirmMembership: z.boolean().refine((bool) => bool == true),
+  })
+  .or(
+    z.object({
+      isCreation: z.literal(true),
+      name: z.string().min(1),
+      description: rte,
+      whatDataMessage: rte,
+      creationConfirmMembership: z.boolean().refine((bool) => bool == true),
+    }),
+  );
+
+type JoinOrganizationFormType = z.infer<typeof joinOrganizationSchema>;
+
+function JoinOrganizationModal({
+  isShow,
+  setIsShow,
+}: {
+  isShow: boolean;
+  setIsShow: Dispatch<SetStateAction<boolean>>;
+}) {
+  const [errorMessage, setErrorMessage] = useState("");
+  const form = useForm<JoinOrganizationFormType>({
+    mode: "onChange",
+    resolver: zodResolver(joinOrganizationSchema),
+    defaultValues: {
+      isCreation: false,
+      message: "",
+      confirmMembership: false,
+    },
+  });
+
+  const onboardMutation = api.user.onboard.useMutation({
+    // TODO: reset is not working
+    // TODO: toast is hiding too fast
```

Comment:
Were these TODOS done?

Reply:
For future reference only, more like a FIXME